File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/CommandBuilders/TFTPErrorCommandBuilder.java
package bgu.spl171.net.impl.TFTP.CommandBuilders;

import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.Responses.TFTPError;
import bgu.spl171.net.impl.TFTP.Commands.WriteRequest;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.ShortEncoderDecoder;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.StringEncoderDecoder;

public class TFTPErrorCommandBuilder implements CommandBuilder {
    Short errorCode = null;
    ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();
    StringEncoderDecoder stringEncoderDecoder = new StringEncoderDecoder();

    @Override
    public Command decodeNextByte(byte nextByte) {
        if (errorCode == null) {
            errorCode = shortEncoderDecoder.decodeNextByte(nextByte);
        } else {
            String errorMsg = stringEncoderDecoder.decodeNextByte(nextByte);
            if (errorMsg != null) {
                return new TFTPError(errorCode, errorMsg);
            }
        }
        return null;
    }
}

File: java to c++ builders/TFTPErrorCommandBuilder.cpp
#include "TFTPErrorCommandBuilder.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandBuilders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::Responses::TFTPError;
						using bgu::spl171::net::impl::TFTP::Commands::WriteRequest;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::ShortEncoderDecoder;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::StringEncoderDecoder;

						Command *TFTPErrorCommandBuilder::decodeNextByte(char nextByte)
						{
							if (!errorCode)
							{
								errorCode = shortEncoderDecoder->decodeNextByte(nextByte);
							}
							else
							{
								std::wstring errorMsg = stringEncoderDecoder->decodeNextByte(nextByte);
								if (errorMsg != L"")
								{
									return new TFTPError(errorCode, errorMsg);
								}
							}
							return nullptr;
						}
					}
				}
			}
		}
	}
}
